File: golang/338. Counting Bits.go
package golang

/** 
* 从1开始，每个数子的二进制位，等于不算最低位的二进制1个数+奇数偶数判断最低位1的个数。
*/
func countBits(num int) []int {
    res := []int{0}
    for i := 1; i <= num; i++ {
        res = append(res, res[i>>1] + i%2)
    }
    return res
}

File: js/338. Counting Bits.js
/**
 * @param {number} num
 * @return {number[]}
 */

var countBits = function(num) {
    var res = [0], cnt = 0
    while (cnt < num) {
        length = res.length
        for (var i=0; i<length && cnt<num; ++i, ++cnt) {
            res.push(res[i]+1)
        }
    }
    return res
};

/**
 * If number is 0, =>0
 * If number is 1, =>1
 * If number is 2, =>1
 * If number is 3, =>2
 * ...
 * The rule is  [1] = [0] + 1
 *              [2, 3] = [0, 1] + 1
 * 
 * Also, we can judge it by odd/even...(golang)
 */